lib/scarpe/download.rb

```diff
@@ -4,44 +4,124 @@
+          handle_response(wrapped_response, save, styles)
+        rescue Net::HTTPError, Net::OpenTimeout, Net::ReadTimeout => e
+          handle_error(e, logger)
+        rescue StandardError => e
+          handle_error(e.message, logger) # Pass the error message as a string
         end
-      rescue StandardError => e
-        Scarpe.error("Error occurred while downloading: #{e.message}")
       end
     end
-  end
 
-  private
+    private
+
+    def perform_request(uri, method, styles)
+      port = uri.port || (uri.scheme == "https" ? 443 : 80)
+      http = Net::HTTP.start(uri.host, port, use_ssl: uri.scheme == "https", verify_mode: OpenSSL::SSL::VERIFY_NONE)
+
+      request = Net::HTTP.const_get(method.capitalize).new(uri.request_uri)
+      apply_styles(request, styles)
+
+      http.request(request)
+    end
+
+    def apply_styles(request, styles)
+      headers = styles[:headers]
+      request["Content-Type"] = headers["Content-Type"] if headers && headers["Content-Type"]
+      request.body = styles[:body] if styles[:body]
+      # Add more custom styles to the request as needed
+    end
+
+    def handle_response(response, save, styles)
+      case response.response.code.to_i
+      when 200..299
+        content = response.body
+        # headers = response.headers # Access headers directly
+        if save
+          save_content(content, save)
+          parse_rss(content) # Parse the downloaded XML content
+        else
+          handle_finish_event(response) # Pass response and headers to handle_finish_event
+        end
+      else
+        handle_failure(response.response.code)
+      end
+    end
+
+    def save_content(content, file_name)
+      file = nil
+      begin
+        file = File.open(file_name, "w")
+        file.write(content)
+      rescue Errno::EACCES, Errno::ENOENT => e
+        raise FileError, "File error occurred: #{e.message}"
+      else
+        @block&.call
+      ensure
+        file&.close
+      end
+    end
 
-  def html_get_title(content)
-    doc = Nokogiri::HTML(content)
+    def handle_finish_event(response)
+      passed = Struct.new(:response).new(response) # Create a new Struct with response attribute
+      @block&.call(passed) # Pass it ok
+    end
 
-    title = doc.at_css("title")&.text&.strip || ""
+    def handle_failure(code, logger)
+      logger.error("Failed to download content. Response code: #{code}")
+    end
 
-    # headings = doc.css("h1").map(&:text)
+    def handle_error(error, logger)
+      logger.error("An error occurred while downloading: #{error}")
+    end
 
-    title
+    def parse_rss(content)
+      doc = Nokogiri::XML(content)
+      items = doc.xpath("//item")
+      items.each do |item|
+        title = item.xpath("title").text
+        link = item.xpath("link").text
+        description = item.xpath("description").text
+        # Process the extracted data as needed
+        puts "Title: #{title}"
```

Comment:
Are the puts statements going to stay?

Reply:
will remove actually i think we dont need tht parse method anymore as i modified example but keeping the method , removing puts

Reply:
removed

Reply:
Thanks!